### PR TITLE
FINERACT-1561 Client and Loan APIs enhancements

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

In attention to the next tickets:

[Apache Fineract JIRA ticket 1559](https://issues.apache.org/jira/browse/FINERACT-1559)
[Apache Fineract JIRA ticket 1560](https://issues.apache.org/jira/browse/FINERACT-1560)
[Apache Fineract JIRA ticket 1561](https://issues.apache.org/jira/browse/FINERACT-1561).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/api-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java [hunk 1]
```diff
@@ -1535,6 +1549,12 @@ private String getAddSql(final List<ResultsetColumnHeaderData> columnHeaders, fi
                 columnName = sqlGenerator.escape(key);
                 insertColumns += ", " + columnName;
                 selectColumns += "," + pValueWrite + " as " + columnName;
+            } else {
+                if (key.equalsIgnoreCase("created_at") || key.equalsIgnoreCase("updated_at")) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: Could we extract the created_at and updated_at strings into constants so they're not everywhere in the datatables codebase as magic strings?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java [hunk 2]
```diff
@@ -639,6 +640,12 @@ public String retrieveLoan(@PathParam("loanId") @Parameter(description = "loanId
                 linkedAccount = this.accountAssociationsReadPlatformService.retriveLoanLinkedAssociation(loanId);
             }
 
+            if (associationParameters.contains("collection")) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java]: This string is still magical, could we extract it into a constant?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 3]
```diff
@@ -2358,4 +2359,56 @@ public Integer retrieveNumberOfActiveLoans() {
...
+        CollectionDataMapper(DatabaseSpecificSQLGenerator sqlGenerator) {
+            this.sqlGenerator = sqlGenerator;
+        }
+
+        public String schema() {
+            StringBuilder sqlBuilder = new StringBuilder();
+
+            sqlBuilder.append(
+                    "l.id as loanId, coalesce((l.approved_principal - l.principal_disbursed_derived), 0) as availableDisbursementAmount, ");
+            sqlBuilder.append("datediff(" + sqlGenerator.currentDate() + ", laa.overdue_since_date_derived) as pastDueDays, ");
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: datediff problem is still present.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 4]
```diff
@@ -2358,4 +2359,56 @@ public Integer retrieveNumberOfActiveLoans() {
...
+            sqlBuilder.append(
+                    "(select coalesce(min(lrs.duedate), null) as duedate from m_loan_repayment_schedule lrs where lrs.loan_id=1 and lrs.completed_derived is false and lrs.duedate >= "
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: loan_id=1 still present.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 5]
```diff
@@ -2358,4 +2359,56 @@ public Integer retrieveNumberOfActiveLoans() {
...
+                            + sqlGenerator.currentDate() + ") as nextPaymentDueDate, ");
+            sqlBuilder.append("datediff(" + sqlGenerator.currentDate() + ", laa.overdue_since_date_derived) as delinquentDays, ");
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: datediff too

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java [hunk 6]
```diff
@@ -2358,4 +2359,56 @@ public Integer retrieveNumberOfActiveLoans() {
...
+            sqlBuilder.append(
+                    sqlGenerator.currentDate() + " as delinquentDate, coalesce(laa.total_overdue_derived, 0) as delinquentAmount, ");
+            sqlBuilder.append("lre.transactionDate as lastPaymentDate, coalesce(lre.amount, 0) as lastPaymentAmount ");
+            sqlBuilder.append("from m_loan l inner join m_loan_arrears_aging laa on laa.loan_id = l.id ");
+            sqlBuilder.append(
+                    "left join (select lt.loan_id, lt.transaction_date as transactionDate, lt.amount as amount from m_loan_transaction lt where lt.is_reversed = 0 and lt.transaction_type_enum=2 order by lt.transaction_date desc limit 1) lre on lre.loan_id = l.id ");
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: is_reversed = 0 is going to fail on PG

**Review by galovics — CHANGES_REQUESTED** [review]
Still things to be fixed, I commented some of them once more but please check the original PR where I left the initial comments.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: Solved

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/api/LoansApiResource.java]: Solved

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: Solved

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java (hunk 4 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: Solved

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java (hunk 5 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanReadPlatformServiceImpl.java]: Solved

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/DataTableApiConstant.java [hunk 7]
```diff
@@ -34,4 +34,25 @@ private DataTableApiConstant() {
     public static final String localParamName = "locale";
     public static final String DATATABLE_RESOURCE_NAME = "dataTables";
 
+    public static final String createdAtFieldName = "created_at";
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/DataTableApiConstant.java]: This naming is not comforming the Java styling standards.
`
The names of variables declared class constants and of ANSI constants should be all uppercase with words separated by underscores ("_"). (ANSI constants should be avoided, for ease of debugging.)` - https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java [hunk 8]
```diff
@@ -1637,6 +1658,10 @@ private String getUpdateSql(List<ResultsetColumnHeaderData> columnHeaders, final
                     }
                 }
                 sql += sqlGenerator.escape(key) + " = " + pValueWrite;
+            } else {
+                if (key.equalsIgnoreCase("updated_at")) {
```
Comment (galovics) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: Magic string here.

**Review by galovics — CHANGES_REQUESTED** [review]
Minor changes still required. Good progress though, thanks @josehernandezfintecheandomx

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/DataTableApiConstant.java (hunk 7 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/api/DataTableApiConstant.java]: Solved

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java (hunk 8 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadWriteNonCoreDataServiceImpl.java]: Solved

**Review by galovics — APPROVED** [review]
Looks good.

**Comment by josehernandezfintecheandomx** [comment]
Included in other PR https://github.com/apache/fineract/pull/2279

**Pull request closed without merging**